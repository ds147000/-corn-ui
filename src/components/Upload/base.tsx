import React, { useEffect, useMemo, useRef } from 'react'
import { View } from '@tarojs/components'
import classNames from 'classnames'
import Image from '../Image'
import previewImage from '../PreviewImage'
import Icon from '../Icon'

export declare namespace Upload {
  type Media = {
    mediaId: number
    content: string
  }

  /** 自定义上传，必须返回上传后的结果 */
  type handleUpload = (file: File) => Promise<Media>

  /** 上传完成回调函数 */
  type onChange = (imgs: Media[]) => void

  type layout = 'square' | 'row'

  type onRemove = (img: Media, index: number) => void

  interface props {
    /** 图片项类名 */
    itemClassName?: string
    /** 最多数量 */
    count?: number
    /** 布局类型 */
    layout?: layout
    /** 自定义渲染上传按钮 */
    btn?: React.ReactNode
    /** 已经上传文件对象 */
    list?: Media[]
    /** 上传项删除事件 */
    onRemove?: onRemove
  }
}

export const DEFAULT_COUNT = 9

export const UploadBase: React.FC<Upload.props> = ({
  itemClassName,
  count = DEFAULT_COUNT,
  layout = 'row',
  btn,
  list,
  onRemove
}) => {
  const El = useRef<HTMLDivElement>()
  const isShowBtn = !list || list.length < count

  const _class = useMemo(() => classNames(
    'xrk-upload',
    'xrk-upload-' + layout
  ), [ layout ])

  const _itemClass = useMemo(() => classNames(
    'xrk-upload-item',
    itemClassName
  ), [ itemClassName ])

  const onPreview = (index: number): void => {
    const _list = list as Upload.Media[]
    previewImage({
      current: _list[index].content,
      urls: _list?.map((item) => item.content)
    })
  }

  useEffect(() => {
    if (layout !== 'row') return

    // #if _APP === 'weapp'

    // #else
    try {
      El.current?.scrollTo?.({ left: El.current.scrollWidth, top: 0, behavior: 'smooth' })
    } catch (error) {
      El.current?.scrollTo?.(El.current.scrollWidth, 0)
    }
    // #endif
  }, [ list?.length, layout ])

  return (
    <View className={_class} ref={El} >
      {list?.map((item, key) => (
        <View key={item.mediaId + item.content} className={_itemClass} >
          <View
            data-testid="upload-item"
            className="xrk-upload-remove xrk-f xrk-ac xrk-jc"
            onClick={(): void => onRemove?.(item, key)}
          >
            <Icon name="delete" data-testid="upload-remove" />
          </View>
          <Image
            className="xrk-upload-img"
            src={item.content}
            onClick={(): void => onPreview(key)}
            lazyLoad
            mode="aspectFill"
          />
        </View>
      ))}
      {isShowBtn && btn}
    </View>
  )
}

