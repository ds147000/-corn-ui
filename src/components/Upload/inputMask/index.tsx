/* eslint-disable import/export */
/* eslint-disable import/first */

// #if _APP === 'weapp'
import { InputMask } from './index.taro'

export default InputMask
// #else
import { InputMask } from './index.h5'

export default InputMask
// #endif
