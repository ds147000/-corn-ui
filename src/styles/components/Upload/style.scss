.xrk-upload {
  display: flex;
  align-items: center;

  &-row {
    overflow-y: auto;
    height: 160px;
  }

  &-square {
    flex-wrap: wrap;
    min-height: 160px;

    .xrk-upload-item, .xrk-upload-btn {
      margin-bottom: 18px;
    }
  }

  &-item {
    width: 160px;
    height: 160px;
    min-width: 160px;
    min-height: 160px;
    border-radius: 30px;
    position: relative;
    margin-right: 18px;
    overflow: hidden;
  }

  &-img {
    display: block;
    width: 160px;
    height: 160px;
    border-radius: 30px;
  }

  &-remove {
    position: absolute;
    top: 0px;
    right: 0px;
    width: 45px;
    height: 45px;
    background: rgba(0, 0, 0, 0.5);
    border-radius: 0px 30px 0px 30px;
    overflow: hidden;
    z-index: 5;
    color: #fff;

    .xrk-icon {
      font-size: 25px;
    }
  }

  &-btn {
    width: 160px;
    height: 160px;
    min-width: 160px;
    min-height: 160px;
    background: $uploadColor;
    border-radius: 30px;
    position: relative;

    .xrk-icon {
      font-size: 48px;
      color: $icon1;
    }
  }

  &-mask {
    position: absolute;
    width: 100%;
    height: 100%;
    opacity: 0 !important;
    border: none;
    outline: none;
  }
}
