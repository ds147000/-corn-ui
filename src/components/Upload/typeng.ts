import { Upload } from './base'
import { InputMaskProps } from './inputMask/typeing'

export type HandelUpload = (file: FileList) => Promise<Upload.Media[]>

export interface UploadProps {
  /** 表单名称 */
  name?: string
  /** 图片项类名 */
  itemClassName?: string
  /** 是否开启多选 */
  multiple?: boolean
  /** 最多数量 */
  count?: number
  /** 布局类型 */
  layout?: Upload.layout
  /** 已经上传文件对象 */
  list?: Upload.Media[]
  /** 文件类型 */
  type?: 'video' | 'image' | 'all'
  /** 自定义按钮渲染 */
  btn?(props: InputMaskProps): JSX.Element
  /** 上传前触发的钩子，返回true终止上传。参数 file:当前选择文件，list: 当前已经上传的列表 */
  beforUpload?(file: FileList, list: Upload.Media[]): boolean
  /** 单独设置处理上传方法 */
  handelUpload?: HandelUpload
}
