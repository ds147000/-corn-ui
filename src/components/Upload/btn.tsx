import { View } from '@tarojs/components'
import React from 'react'
import Icon from '../Icon'
import InputMask from './inputMask'
import { InputMaskProps } from './inputMask/typeing'

export const UploadDefaultBtn: React.FC<InputMaskProps> = (props) => {
  return (
    <View className="xrk-upload-btn xrk-f xrk-ac xrk-jc" data-testid="upload-btn">
      <InputMask {...props} />
      <Icon name="img" />
    </View>
  )
}
