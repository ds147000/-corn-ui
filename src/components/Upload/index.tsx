/* eslint-disable no-magic-numbers */
import React from 'react'
import { DEFAULT_COUNT, Upload, UploadBase } from './base'
import { UploadDefaultBtn } from './btn'
import { HandelUpload, UploadProps } from './typeng'
import Toast from '../Toast'
import { InputMaskProps } from './inputMask/typeing'

interface UploadState {
  list: Upload.Media[]
}

const OPEN_MUTIPLE_COUNT = 2

class UploadComps extends React.Component<UploadProps, UploadState> {
  static handelUpload: HandelUpload
  /** 获取处理上传方法 */
  get handelUpload(): HandelUpload {
    return this.props.handelUpload || UploadComps.handelUpload
  }

  /** 当前文件类型 */
  get fileType(): 'all' | 'video' | 'image' {
    return this.props.type || 'all'
  }

  /** 是否开启多选 */
  get isMultiple(): boolean {
    return this.props.multiple ?? this.state.list.length - (this.props.count || DEFAULT_COUNT) < OPEN_MUTIPLE_COUNT
  }

  get count(): number {
    return this.props.count || DEFAULT_COUNT
  }

  state: UploadState = {
    list: []
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
  onChange = (file: any): void => {
    if (this.props.beforUpload?.(file, this.state.list)) return
    if (this.checkListLen(file)) return

    // #if _APP === 'weapp'

    // #else
    this.onChangeOfH5(file)
    // #endif
  }

  onChangeOfH5 = async (file: FileList): Promise<void> => {
    try {
      const list = await this.handelUpload(file)
      this.setState({ list: [ ...this.state.list, ...list ] })
    } catch (error) {
      Toast.show(error.message)
    }
  }

  onRemove = (__: Upload.Media, index: number): void => {
    this.setState({
      list: this.state.list.filter((_, key) => key !== index)
    })
  }

  private checkListLen = (file: FileList): boolean => {
    const len = this.state.list.length + file.length
    if (len > this.count) {
      Toast.show('数量超出限制，请重新选择')
      return true
    }

    return false
  }

  render(): JSX.Element {
    const btnProps: InputMaskProps = {
      type: this.fileType,
      onChange: this.onChange,
      name: this.props.name,
      multiple: this.isMultiple
    }

    return (
      <UploadBase
        {...this.props}
        list={this.state.list}
        onRemove={this.onRemove}
        btn={this.props.btn ? this.props.btn(btnProps) : <UploadDefaultBtn {...btnProps} />}
      />
    )
  }
}

UploadComps.handelUpload = async (file: FileList): Promise<Upload.Media[]> => {
  const list: Upload.Media[] = []
  for(let i = 0; i < file.length; i++) {
    list.push({ mediaId: Math.random(), content: URL.createObjectURL(file[i]) })
  }
  return Promise.resolve(list)
}

export default UploadComps
