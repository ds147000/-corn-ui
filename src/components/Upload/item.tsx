// #if _APP === 'weapp'
import Taro from '@tarojs/taro'
// #endif
import classNames from 'classnames'
import { View } from '@tarojs/components'
import React, { useMemo } from 'react'
import { Upload } from './base'
import Image from '../Image'
import Icon from '../Icon'


interface UploadItemProps {
  onPreview(item: Upload.Media): void
  onRemove?: Upload.onRemove
  /** 图片项类名 */
  className?: string
  item: Upload.Media
}

export const UploadItem: React.FC<UploadItemProps> = ({
  onPreview,
  onRemove,
  className,
  item
}) => {
  const _itemClass = useMemo(() => classNames(
    'corn-upload-item',
    className
  ), [ className ])

  const status = useMemo(() => {
    switch(item.status) {
      case 'loading':
        return <View className="corn-upload-loading" />

      case 'error':
        return <View className="corn-upload-error corn-f corn-ac corn-jc" >上传失败</View>

      default:
        return null
    }
  }, [ item.status ])

  const _onRemove = (): void => {
    onRemove?.(item)
  }

  const _onPreview = (): void => {
    onPreview(item)
  }

  return (
    <View className={_itemClass} data-testid="upload-item" >
      <View
        className="corn-upload-remove corn-f corn-ac corn-jc"
        onClick={_onRemove}
        data-testid="upload-remove"
      >
        <Icon name="delete"  />
      </View>
      <Image
        className="corn-upload-img"
        src={item.content}
        onClick={_onPreview}
        lazyLoad
        mode="aspectFill"
        // #if _APP === 'weapp'
        // eslint-disable-next-line no-magic-numbers
        style={{ width: Taro.pxTransform(160), height: Taro.pxTransform(160) }}
        // #endif
      />
      {status}
    </View>
  )
}
