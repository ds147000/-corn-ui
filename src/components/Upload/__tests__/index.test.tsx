import { render, waitFor } from '@testing-library/react'
import React from 'react'
import Upload from '../index'
import UserEvent from '@testing-library/user-event'

jest.mock('../inputMask')
jest.mock('../../PreviewImage')
jest.mock('../../Toast')

describe('Uplaod.h5', () => {

  test('快照', async () => {
    const screen = render(<Upload />)
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('渲染自定义按钮', async () => {
    const screen = render(<Upload btn={() => <div>upload</div>} />)
    screen.getByText('upload')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('上传', async () => {
    const handelUpload = jest.fn()
    const screen = render(<Upload handelUpload={handelUpload} />)
    const file = new File([], '123.jpg', { type: 'image/jpg' })
    UserEvent.upload(screen.getByTestId('uplaod-btn'), file)
    await waitFor(() => expect(handelUpload).toHaveBeenCalledTimes(1))
  })

  test('上传成功', async () => {
    let a = 1
    const handelUpload = jest.fn((file: FileList) => Promise.resolve([{ mediaId: a++, content: '123.jpg' }]))

    const screen = render(<Upload handelUpload={handelUpload} />)
    const file = new File([], '123.jpg', { type: 'image/jpg' })
    const file2 = new File([], '123.jpg', { type: 'image/jpg' })
    const file3 = new File([], '123.jpg', { type: 'image/jpg' })
    UserEvent.upload(screen.getByTestId('uplaod-btn'), file)
    await waitFor(async () => expect(await (await screen.findAllByTestId('upload-item')).length).toBe(1))
    UserEvent.upload(screen.getByTestId('uplaod-btn'), file2)
    UserEvent.upload(screen.getByTestId('uplaod-btn'), file3)
    await waitFor(async () => expect(await (await screen.findAllByTestId('upload-item')).length).toBe(3))
  })

  test('上传失败', async () => {
    const handelUpload = jest.fn((file: FileList) => Promise.reject(new Error('上传失败')))

    const screen = render(<Upload handelUpload={handelUpload} />)
    const file = new File([], '123.jpg')
    UserEvent.upload(screen.getByTestId('uplaod-btn'), file)
  })

  test('指定全局上传成功', async () => {
    let a = 1
    const handelUpload = jest.fn((file: FileList) => Promise.resolve([{ mediaId: a++, content: '123.jpg' }]))
    Upload.handelUpload = handelUpload

    const screen = render(<Upload />)
    const file = new File([], '123.jpg', { type: 'image/jpg' })
    const file2 = new File([], '123.jpg', { type: 'image/jpg' })
    const file3 = new File([], '123.jpg', { type: 'image/jpg' })
    UserEvent.upload(screen.getByTestId('uplaod-btn'), file)
    await waitFor(async () => expect(await (await screen.findAllByTestId('upload-item')).length).toBe(1))
    UserEvent.upload(screen.getByTestId('uplaod-btn'), file2)
    UserEvent.upload(screen.getByTestId('uplaod-btn'), file3)
    await waitFor(async () => expect(await (await screen.findAllByTestId('upload-item')).length).toBe(3))
  })
})
