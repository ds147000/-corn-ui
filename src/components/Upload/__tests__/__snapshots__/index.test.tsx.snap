// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Uplaod.h5 快照 1`] = `
<div>
  <div
    class="xrk-upload xrk-upload-row"
  >
    <div
      class="xrk-upload-btn xrk-f xrk-ac xrk-jc"
      data-testid="upload-btn"
    >
      <input
        accept="video/*,image/*"
        class="xrk-upload-mask"
        data-testid="uplaod-btn"
        multiple=""
        type="file"
      />
      <i
        class="xrk-icon xrk-i-img"
      />
    </div>
  </div>
</div>
`;

exports[`Uplaod.h5 渲染自定义按钮 1`] = `
<div>
  <div
    class="xrk-upload xrk-upload-row"
  >
    <div>
      upload
    </div>
  </div>
</div>
`;
